File: android/src/main/java/com/kustomer/chatreactnative/providers/RNKustomerChatProvider.kt
package com.kustomer.chatreactnative.providers

import com.facebook.react.bridge.*
import com.kustomer.chatreactnative.*
import com.kustomer.core.KustomerCore
import com.kustomer.core.exception.KusNotInitializedException
import com.kustomer.core.models.KusResult
import com.kustomer.core.models.KusChatSetting
import com.kustomer.core.models.chat.*
import com.kustomer.core.models.pubnub.KusPresenceEvent
import com.kustomer.ui.Kustomer
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch

class RNKustomerChatProvider(reactContext: ReactApplicationContext) :
  RNKustomerChatProviderSpec(reactContext) {
  companion object {
    const val NAME = "RNKustomerChatProvider"
  }
  override fun getName() = NAME

  @ReactMethod
  override fun createConversation(
    newConversationOptions: ReadableMap,
    promise: Promise
  ) {
    CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate).launch {
      val title: String? = newConversationOptions.getString("title")
      val messages: ReadableArray? = newConversationOptions.getArray("messages")
      val messageList: List<String> = ArrayUtil.toArray(messages).asList().map { it as String }

      when (val it: KusResult<Pair<KusConversation, List<KusChatMessage>>> = KustomerCore.getInstance().kusChatProvider().createConversation(
        title, messageList, listOf()
      )) {
        is KusResult.Success -> {
          val (conversation, ) = it.data
          val payload: WritableMap = toWritableKustomerConversation(conversation)
          promise.resolve(payload)
        }
        is KusResult.Error -> promise.reject(
          "$NAME::CREATE_CONVERSATION_ERROR",
          it.exception.localizedMessage, it.exception
        )
        else -> promise.reject(
          "$NAME::CREATE_CONVERSATION_ERROR",
          "Unknown error"
        )
      }
    }
  }

  @ReactMethod
  fun fetchConversations(page: Double, pageSize: Double, promise: Promise) {
    CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate).launch {
      when (val it: KusResult<List<KusConversation>> = KustomerCore.getInstance().kusChatProvider().fetchConversations(page.toInt(), pageSize.toInt())) {
        is KusResult.Success -> {
          val arr: Array<WritableMap> = it.data.map { conversation -> toWritableKustomerConversation(conversation) }.toTypedArray()
          val payload: WritableArray = ArrayUtil.toWritableArray(arr)
          promise.resolve(payload)
        }
        is KusResult.Error -> promise.reject(
          "$NAME::FETCH_CONVERSATIONS_ERROR",
          it.exception.localizedMessage, it.exception
        )
        else -> promise.reject(
          "$NAME::FETCH_CONVERSATIONS_ERROR",
          "Unknown error"
        )
      }
    }
  }

  @ReactMethod
  override fun endConversation(conversationId: String, promise: Promise) {
    CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate).launch {
      when (val it: KusResult<KusConversation> = KustomerCore.getInstance().kusChatProvider().endConversation(conversationId)) {
        is KusResult.Success -> {
          val payload: WritableMap = toWritableKustomerConversation(it.data)
          promise.resolve(payload)
        }
        is KusResult.Error -> promise.reject(
          "$NAME::END_CONVERSATION_ERROR",
          it.exception.localizedMessage, it.exception
        )
        else -> promise.reject(
          "$NAME::END_CONVERSATION_ERROR",
          "Unknown error"
        )
      }
    }
  }

  @ReactMethod
  fun fetchConversation(conversationId: String, promise: Promise) {
    CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate).launch {
      when (val it: KusResult<KusConversation> = KustomerCore.getInstance().kusChatProvider().fetchConversation(conversationId)) {
        is KusResult.Success -> {
          val payload: WritableMap = toWritableKustomerConversation(it.data)
          promise.resolve(payload)
        }
        is KusResult.Error -> promise.reject(
          "$NAME::FETCH_CONVERSATION_ERROR",
          it.exception.localizedMessage, it.exception
        )
        else -> promise.reject(
          "$NAME::FETCH_CONVERSATION_ERROR",
          "Unknown error"
        )
      }
    }
  }

  @ReactMethod
  fun markRead(conversationId: String, messageIds: ReadableArray, promise: Promise) {
    val messageIdsList: List<String> = ArrayUtil.toArray(messageIds).asList().map { it as String }

    CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate).launch {
      KustomerCore.getInstance().kusChatProvider().markRead(conversationId, messageIdsList)
      promise.resolve(null)
    }
  }

  @ReactMethod
  fun fetchChatMessages(conversationId: String, pageSize: Double, startTimeToken: Double, promise: Promise) {
    CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate).launch {
      when (val it: KusResult<List<Any>> = KustomerCore.getInstance().kusChatProvider().fetchChatMessages(conversationId, pageSize.toInt(), startTimeToken.toInt())) {
        is KusResult.Success -> promise.resolve(it.data) // todo
        is KusResult.Error -> promise.reject(
          "$NAME::FETCH_CHAT_MESSAGES_ERROR",
          it.exception.localizedMessage, it.exception
        )
        else -> promise.reject(
          "$NAME::FETCH_CHAT_MESSAGES_ERROR",
          "Unknown error"
        )
      }
    }
  }

  @ReactMethod
  fun sendMessageWithAttachments(
    body: String?,
    messageActionMap: ReadableMap,
//    attachments: List<KusUploadAttachment>?, // TODO: add support for attachments
    conversationId: String,
    isFirstMessage: Boolean,
    promise: Promise
  ) {

    val messageAction: KusMessageAction = toKusMessageAction(messageActionMap)

    CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate).launch {
      when (val it: KusResult<KusChatMessage> = KustomerCore.getInstance().kusChatProvider().sendMessageWithAttachments(
        body, messageAction, listOf(), conversationId, isFirstMessage
      )) {
        is KusResult.Success -> {
          val payload: WritableMap = toWritableKustomerChatMessage(it.data)
          promise.resolve(payload)
        }
        is KusResult.Error -> promise.reject(
          "$NAME::SEND_MESSAGE_WITH_ATTACHMENTS_ERROR",
          it.exception.localizedMessage, it.exception
        )
        else -> promise.reject(
          "$NAME::SEND_MESSAGE_WITH_ATTACHMENTS_ERROR",
          "Unknown error"
        )
      }
    }
  }


  @ReactMethod
  override fun sendTypingStatus(conversationId: String, typingStatus: String, promise: Promise) {
    val stauts = toKusTypingStatus(typingStatus)

    CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate).launch {
      when (val it: KusResult<Boolean> = KustomerCore.getInstance().kusChatProvider().sendTypingStatus(conversationId, stauts)) {
        is KusResult.Success -> promise.resolve(it.data)
        is KusResult.Error -> promise.reject(
          "$NAME::SEND_TYPING_STATUS_ERROR",
          it.exception.localizedMessage, it.exception
        )
        else -> promise.reject(
          "$NAME::SEND_TYPING_STATUS_ERROR",
          "Unknown error"
        )
      }
    }
  }

  @ReactMethod
  override fun isChatAvailable(promise: Promise) = try {
    Kustomer.getInstance().isChatAvailable {
      when (it) {
        is KusResult.Success -> promise.resolve(it.data)
        is KusResult.Error -> promise.reject(
          "$NAME::IS_CHAT_AVAILABLE_ERROR", it.exception.localizedMessage
        )
        else -> promise.reject("$NAME::IS_CHAT_AVAILABLE_ERROR", it.toString())
      }
    }
  } catch (e: KusNotInitializedException) {
    promise.reject("$NAME::IS_CHAT_AVAILABLE_ERROR", e)
  }

  @ReactMethod
  override fun getChatSettings(promise: Promise) {
    CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate).launch {
      when (val it: KusResult<KusChatSetting> = KustomerCore.getInstance().kusChatProvider().getChatSettings()) {
        is KusResult.Success -> {
          val payload: WritableMap = toWritableKustomerChatSetting(it.data)
          promise.resolve(payload)
        }
        is KusResult.Error -> promise.reject(
          "$NAME::GET_CHAT_SETTINGS_ERRROR",
          it.exception.localizedMessage, it.exception
        )
        else -> promise.reject(
          "$NAME::GET_CHAT_SETTINGS_ERRROR",
          "Unknown error"
        )
      }
    }
  }

  @ReactMethod
  override fun overrideBusinessSchedule(scheduleId: String, promise: Promise) = try {
    Kustomer.getInstance().overrideBusinessSchedule(scheduleId) {
      when (it) {
        is KusResult.Success -> promise.resolve(it.data)
        is KusResult.Error -> promise.reject(
          "$NAME::OVERRIDE_BUSINESS_SCHEDULE_ERROR", it.exception.localizedMessage
        )
        else -> promise.reject("$NAME::OVERRIDE_BUSINESS_SCHEDULE_ERROR", it.toString())
      }
    }
  } catch (e: KusNotInitializedException) {
    promise.reject("$NAME::OVERRIDE_BUSINESS_SCHEDULE_ERROR", e)
  }

  @ReactMethod
  override fun changeActiveAssistant(assistantId: String, promise: Promise) {
    val activeAssistant = KusActiveAssistant.WithId(assistantId)

    CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate).launch {
      when (val it: KusResult<KusAssistant>? = KustomerCore.getInstance().kusChatProvider().changeActiveAssistant(activeAssistant)) {
        is KusResult.Success -> {
          val payload: WritableMap = toWritableKustomerAssistant(it.data)
          promise.resolve(payload)
        }
        is KusResult.Error -> promise.reject(
          "$NAME::CHANGE_ACTIVE_ASSISTANT_ERROR",
          it.exception.localizedMessage, it.exception
        )
        else -> promise.reject(
          "$NAME::CHANGE_ACTIVE_ASSISTANT_ERROR",
          "Unknown error"
        )
      }
    }
  }

  @ReactMethod
  override fun init(promise: Promise) {
    // todo
  }

  @ReactMethod
  override fun setPresenceActivity(activity: String)  {
    val presence = toKusPresenceEvent(activity)
    KustomerCore.getInstance().kusChatProvider().sendPresenceActivity(presence)
  }

  @ReactMethod
  override fun describeCustomer(attributes: ReadableMap, promise: Promise) {
    val customer = toKusCustomerDescribeAttributes(attributes)

    CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate).launch {
      when (val it: KusResult<Boolean> = KustomerCore.getInstance().kusChatProvider().describeCustomer(customer)) {
        is KusResult.Success -> {
          promise.resolve(it.data)
        }
        is KusResult.Error -> promise.reject(
          "$NAME::DESCRIBE_CUSTOMER_ERROR",
          it.exception.localizedMessage, it.exception
        )
        else -> promise.reject(
          "$NAME::DESCRIBE_CUSTOMER_ERROR",
          "Unknown error"
        )
      }
    }
  }

  @ReactMethod
  override fun getInitialMessages(promise: Promise) {
    CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate).launch {
      when (val it: KusResult<List<KusChatMessage>> = KustomerCore.getInstance().kusChatProvider().getInitialMessages()) {
        is KusResult.Success -> {
          val arr: Array<WritableMap> = it.data.map { message -> toWritableKustomerChatMessage(message) }.toTypedArray()
          val payload: WritableArray = ArrayUtil.toWritableArray(arr)
          promise.resolve(payload)
        }
        is KusResult.Error -> promise.reject(
          "$NAME::GET_INITIAL_MESSAGES_ERROR",
          it.exception.localizedMessage, it.exception
        )
        else -> promise.reject(
          "$NAME::GET_INITIAL_MESSAGES_ERROR",
          "Unknown error"
        )
      }
    }
  }

  @ReactMethod
  override fun describeConversation(conversationId: String, attributes: ReadableMap, promise: Promise) {
    val attrMap: Map<String, Any> = MapUtil.toMap(attributes)

    CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate).launch {
      when (val it: KusResult<Boolean> = KustomerCore.getInstance().kusChatProvider().describeConversation(conversationId, attrMap)) {
        is KusResult.Success -> {
          promise.resolve(it.data)
        }
        is KusResult.Error -> promise.reject(
          "$NAME::DESCRIBE_CONVERSATION_ERROR",
          it.exception.localizedMessage, it.exception
        )
        else -> promise.reject(
          "$NAME::DESCRIBE_CONVERSATION_ERROR",
          "Unknown error"
        )
      }
    }
  }

  @ReactMethod
  override fun getLastReadTimeToken(conversationId: String, promise: Promise) {
    // todo
  }

  @ReactMethod
  override fun sendDeflection(conversationId: String, deflection: ReadableMap, promise: Promise) {
    // todo
  }

  @ReactMethod
  override fun sendChatMessage(conversationId: String, message:String, promise: Promise) {
    val messageAction = null
    val attachments = null

    CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate).launch {
      when (val it: KusResult<KusChatMessage> = KustomerCore.getInstance().kusChatProvider().sendMessageWithAttachments(
        message,
        messageAction,
        attachments,
        conversationId
      )) {
        is KusResult.Success -> {
          val payload = toWritableKustomerChatMessage(it.data)
          promise.resolve(payload)
        }
        is KusResult.Error -> promise.reject(
          "$NAME::SEND_CHAT_MESSAGE_ERROR",
          it.exception.localizedMessage, it.exception
        )
        else -> promise.reject(
          "$NAME::SEND_CHAT_MESSAGE_ERROR",
          "Unknown error"
        )
      }
    }
  }

  @ReactMethod
  override fun sendChatImage(conversationId: String, imageUri: String, promise: Promise) {
    // todo
  }

  @ReactMethod
  override fun sendChatFile(conversationId: String, fileUri: String, promise: Promise) {
    // todo
  }

  @ReactMethod
  override fun sendChatAssistantMessage(conversationId: String, assistantId: String, message:String, promise: Promise) {
    // todo
  }

  @ReactMethod
  override fun sendChatAssistantAction(conversationId: String, action: ReadableMap, additionalParams: ReadableMap, promise: Promise) {
    // todo
  }

  @ReactMethod
  override fun sendChatAssistantMllAction(conversationId: String, mllNode: ReadableMap, additionalParams: ReadableMap, promise: Promise) {
    // todo
  }

  @ReactMethod
  override fun submitSatisfactionForm(conversationId: String, conversationSatisfactionId: String, response: ReadableMap, promise: Promise) {
    val satisfactionPostBody: KusSatisfactionNetworkPostBody = toKusSatisfactionNetworkPostBody(response)

    CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate).launch {
      when (val it: KusResult<KusSatisfaction> = KustomerCore.getInstance().kusChatProvider().submitSatisfactionForm(
        conversationSatisfactionId,
        satisfactionPostBody,
        conversationId
      )) {
        is KusResult.Success -> {
          val satisfaction = toWritableKustomerSatisfaction(it.data)
          promise.resolve(satisfaction)
        }
        is KusResult.Error -> promise.reject(
          "$NAME::SUBMIT_SATISFACTION_FORM_ERROR",
          it.exception.localizedMessage, it.exception
        )
        else -> promise.reject(
          "$NAME::SUBMIT_SATISFACTION_FORM_ERROR",
          "Unknown error"
        )
      }
    }
  }

  @ReactMethod
  override fun getChatMessages(conversationId: String, promise: Promise) {
    // todo
  }

  @ReactMethod
  override fun getHistoricChatMessages(conversationId: String, page: Double, pageSize: Double, promise: Promise) {
    // todo
  }

  @ReactMethod
  override fun getChatMessagesBefore(conversationId: String, timestamp: Double, count: Double, promise: Promise) {
    CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate).launch {
      when (val it: KusResult<List<Any>> = KustomerCore.getInstance().kusChatProvider().getChatMessagesBeforeTimestamp(
        conversationId,
        timestamp.toLong(),
        count.toInt()
      )) {
        is KusResult.Success -> {
          promise.resolve(it.data)
        }
        is KusResult.Error -> promise.reject(
          "$NAME::GET_CHAT_MESSAGES_BEFORE_ERROR",
          it.exception.localizedMessage, it.exception
        )
        else -> promise.reject(
          "$NAME::GET_CHAT_MESSAGES_BEFORE_ERROR",
          "Unknown error"
        )
      }
    }
  }

  @ReactMethod
  override fun getConversation(conversationId: String, promise: Promise) {
    CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate).launch {
      when (val it: KusResult<KusConversation> = KustomerCore.getInstance().kusChatProvider().getConversation(
        conversationId,
      )) {
        is KusResult.Success -> {
          val conversation = toWritableKustomerConversation(it.data)
          promise.resolve(conversation)
        }
        is KusResult.Error -> promise.reject(
          "$NAME::GET_CONVERSATION_ERROR",
          it.exception.localizedMessage, it.exception
        )
        else -> promise.reject(
          "$NAME::GET_CONVERSATION_ERROR",
          "Unknown error"
        )
      }
    }
  }

  @ReactMethod
  override fun getConversations(page: Double, pageSize: Double, promise: Promise) {
    // todo
  }

  @ReactMethod
  override fun getAllConversations(promise: Promise) {
    // todo
  }

  @ReactMethod
  override fun getActiveConversations(promise: Promise) {
    // todo
  }

  @ReactMethod
  override fun getOpenConversationCount(promise: Promise) {
    // todo
  }

  @ReactMethod
  override fun getUnreadCount(promise: Promise) {
    // todo
  }

  @ReactMethod
  override fun fetchSatisfactionFormWithResponse(conversationId: String, promise: Promise) {
    // todo
  }
}

File: chat-react-native/ios/Providers/RNKustomerChatProvider.swift
import KustomerChat
import UIKit

@objcMembers
@objc(RNKustomerChatProvider)
class RNKustomerChatProvider: NSObject {
  static func requiresMainQueueSetup() -> Bool {
    return false
  }

  func constantsToExport() -> [AnyHashable: Any]! {
    return [:]
  }

  // #region Settings
  func isChatAvailable(_ resolve: RCTPromiseResolve<Bool>, rejecter reject: RCTPromiseReject) {
    resolve(Kustomer.chatProvider.isChatAvailable() ?? false)
  }

  func getChatSettings(
    _ resolve: RCTPromiseResolve<[String: Any?]>, rejecter reject: RCTPromiseReject
  ) {
    resolve(Kustomer.chatProvider.getChatSettings().toDictionary())
  }

  func overrideBusinessSchedule(
    _ resolve: @escaping RCTPromiseResolve<[String: Any?]>, rejecter reject: @escaping RCTPromiseReject,
    scheduleId: String
  ) {
    Kustomer.changeBusinessSchedule(scheduleId: scheduleId) { r in
      switch r {
      case .success():
        resolve([:])
        break
      case .failure(let err):
        reject(
          "\("RNKusomterChat")::OVERRIDE_BUSINESS_SCHEDULE_ERROR",
          "Unable to override business schedule: \(err)", err)
        break
      }
    }
  }

  func changeActiveAssistant(
    _ resolve: RCTPromiseResolve<[String: Any?]>, rejecter reject: RCTPromiseReject,
    assistant: NSDictionary
  ) {
      let type = assistant["type"] as! String
      var activeAssistant: KustomerChat.ActiveAssistantOptions? = nil
      switch type {
      case "withId":
          activeAssistant = .withId(assistant["payload"] as! String)
          break;
      default:
          activeAssistant = KustomerChat.ActiveAssistantOptions(rawString: type)
          break;
      }
      KustomerClient.shared.changeActiveAssistant(activeAssistant!)
      resolve([:])
  }
  // #endregion

  // #region meta
  func initinalise(_ resolve: RCTPromiseResolve<Bool>, rejecter reject: RCTPromiseReject) {
      resolve(true) // todo: is this correct?
  }

  func setPresenceActivity(presence: String) {
      Kustomer.chatProvider.sendPresenceActivity(KustomerChat.PresenceType(type: presence))
  }
  func describeCustomer(  // describeCurrentCustomer
    _ resolve: @escaping RCTPromiseResolve<Bool>, rejecter reject: @escaping RCTPromiseReject,
    describeAttributes: NSDictionary
  ) {
      Kustomer.chatProvider.describeCurrentCustomer(phone: describeAttributes["phone"] as? String,
                                             email: describeAttributes["email"] as? String,
                                             phones: describeAttributes["phones"] as? [String],
                                             emails: describeAttributes["emails"] as? [String],
                                             facebook: describeAttributes["facebook"] as? String,
                                             instagram: describeAttributes["instagram"] as? String,
                                             twitter: describeAttributes["twitter"] as? String,
                                             linkedIn: describeAttributes["linkedIn"] as? String,
                                             custom: describeAttributes["custom"] as? [String : Any],
                                               { r in
          switch r {
          case .success():
              resolve(true)
          case .failure(let err):
              reject(
                "\("RNKusomterChat")::DESCRIBE_CUSTOMER_ERROR",
                "Unable to describe customer: \(err)", err)
          }
      })
  }
  // #endregion

  // #region Start/end messaging
  func getInitialMessages(
    _ resolve: RCTPromiseResolve<[String: Any?]>, rejecter reject: RCTPromiseReject
  ) {

  }
  func getInitialMessages(
    _ resolve: RCTPromiseResolve<[String: Any?]>, rejecter reject: RCTPromiseReject,
    assistantId: String?,
    startDialog: String?
  ) {
      reject("\("RNKustomerChatProvider")::NO_IMPL", "Feature not implemented", nil)
  }
  func createConversation(
    _ resolve: RCTPromiseResolve<[String: Any?]>, rejecter reject: RCTPromiseReject,
    newConversationOptions: NSDictionary
  ) {
      reject("\("RNKustomerChatProvider")::NO_IMPL", "Feature not implemented", nil)
  }
  func endConversation(
    _ resolve: RCTPromiseResolve<[String: Any?]>, rejecter reject: RCTPromiseReject,
    conversationId: String
  ) {
      reject("\("RNKustomerChatProvider")::NO_IMPL", "Feature not implemented", nil)
  }
  // #endregion

  // #region Messaging
  func describeConversation(
    _ resolve: RCTPromiseResolve<Void?>, rejecter reject: RCTPromiseReject,
    conversationId: String,
    attributes: NSDictionary
  ) {
      reject("\("RNKustomerChatProvider")::NO_IMPL", "Feature not implemented", nil)
  }
  func getLastReadTimeToken(
    _ conversationId: String, resolve: RCTPromiseResolve<Void?>, rejecter reject: RCTPromiseReject
  ) {
//      resolve(Kustomer.chatProvider.getLastReadTimeToken(for: conversationId))
      reject("\("RNKustomerChatProvider")::NO_IMPL", "Feature not implemented", nil)
  }  // todo: can be done on android through getLastReadTimestamp | getTimeTokenMap
  func sendDeflection(  // sendArticlesClicked on Android
    _ conversationId: String,
    deflection: Any?,
    resolve: RCTPromiseResolve<[String: Any?]>, rejecter reject: RCTPromiseReject
  ) {
      reject(
        "\("RNKusomterChat")::NO_IMPL",
        "Feature not implemented", nil)
//      Kustomer.chatProvider.sendDeflectionTracking(.init(articles: []), conversationId) // todo impl
  }  // todo: iOS doesn't report issues, but android does
  func markRead(
    conversationId: String,
    messageIds: [String]?
  ) {
  }
  func sendTypingStatus(
    conversationId: String,
    typing: Bool = false
  ) {
    Kustomer.chatProvider.sendTypingStatus(for: conversationId, isTyping: typing)
  }
  func sendChatMessage(  // sendMessageWithAttachments on Android
    _ conversationId: String,
    message: String,
    resolve: RCTPromiseResolve<[String: Any?]>, rejecter reject: RCTPromiseReject
  ) {
      reject("\("RNKustomerChatProvider")::NO_IMPL", "Feature not implemented", nil)
  }
  /** Max size for a file: 5MB */
  func sendChatImage(
    _ conversationId: String,
    /* uri of image from camera tmp save, or elsewhere, use Image.resolveAssetSource(foo).uri if needed */
    imageUri: String,
    resolve: RCTPromiseResolve<[String: Any?]>, rejecter reject: RCTPromiseReject
  ) {
      reject("\("RNKustomerChatProvider")::NO_IMPL", "Feature not implemented", nil)
  }
  /** Max size for a file: 5MB */
  func sendChatFile(  // todo: android can send multiple
    _ conversationId: String,
    /* uri of file to send */
    fileUri: String,
    resolve: RCTPromiseResolve<[String: Any?]>, rejecter reject: RCTPromiseReject
  ) {
      reject("\("RNKustomerChatProvider")::NO_IMPL", "Feature not implemented", nil)
  }
  func sendChatAssistantMessage(  // todo: android doesn't distinguish
    _ conversationId: String,
    assistantId: String,
    message: String,
    resolve: RCTPromiseResolve<[String: Any?]>, rejecter reject: RCTPromiseReject
  ) {
      reject("\("RNKustomerChatProvider")::NO_IMPL", "Feature not implemented", nil)
  }
  func sendChatAssistantAction(
    _ conversationId: String,
    action: KUSMessageAction,
    additionalParams: NSDictionary,
    resolve: RCTPromiseResolve<[String: Any?]>, rejecter reject: RCTPromiseReject
  ) {
      reject("\("RNKustomerChatProvider")::NO_IMPL", "Feature not implemented", nil)
  }
  func sendChatAssistantMllAction(
    _ conversationId: String,
    mllNode: KUSMLLNode,
    additionalParams: NSDictionary,
    resolve: RCTPromiseResolve<[String: Any?]>, rejecter reject: RCTPromiseReject
  ) {
      reject("\("RNKustomerChatProvider")::NO_IMPL", "Feature not implemented", nil)
  }
  /**
    * Submits a satisfaction form response for a conversation.
    * @param conversationSatisfactionId Get from `onSatisfactionEventReceived` event
    */
  func submitSatisfactionForm(
    _ conversationId: String,
    conversationSatisfactionId: String,
    response: NSDictionary,
    resolve: RCTPromiseResolve<[String: Any?]>, rejecter reject: RCTPromiseReject
  ) {
      reject("\("RNKustomerChatProvider")::NO_IMPL", "Feature not implemented", nil)
  }
  // #endregion

  // #region Messaging history
  /** Gets the messages for a conversation from PubNub history (up to 30 days). */
  func getChatMessages(
    _ conversationId: String,
    resolve: RCTPromiseResolve<[String: Any?]>, rejecter reject: RCTPromiseReject
  ) {
      reject("\("RNKustomerChatProvider")::NO_IMPL", "Feature not implemented", nil)
  }  // android has paging but noop
  /** Eagerly fetches recent (30 days) chat messages for the conversation */
  // getRecentChatMessages(
  //   conversationId: String
  // ): Promise<KusChatMessage[]>; // todo: android `fetchChatMessages`, page and pageSize are noop
  /** A less optimal way to fetch messages, primarily to be used to go beyond the last 30 days */
  func getHistoricChatMessages(  // todo: iOS is hardcoded page1@100 on historic, can use `getMessages` with `before` and `count` derived from page and pageSize
    _ conversationId: String,
    page: NSNumber,
    pageSize: NSNumber,
    resolve: RCTPromiseResolve<[String: Any?]>, rejecter reject: RCTPromiseReject
  ) {
      reject("\("RNKustomerChatProvider")::NO_IMPL", "Feature not implemented", nil)
  }
  func getChatMessagesBefore(  // todo: android getHistoricChatMessages
    _ conversationId: String,
    timestamp: String,
    count: NSNumber,
    resolve: RCTPromiseResolve<[String: Any?]>, rejecter reject: RCTPromiseReject
  ) {
      reject("\("RNKustomerChatProvider")::NO_IMPL", "Feature not implemented", nil)
  }
  // #endregion

  // #region Conversations
  //? android has fetchX versions, should that be on iOS too?
  func getConversation(
    _ conversationId: String,
   resolve: RCTPromiseResolve<[String: Any?]>, rejecter reject: RCTPromiseReject
  ) {
      reject("\("RNKustomerChatProvider")::NO_IMPL", "Feature not implemented", nil)
  }
  func getConversations(
    _ pageSize: NSNumber,
    page: NSNumber,
    resolve: RCTPromiseResolve<[String: Any?]>, rejecter reject: RCTPromiseReject
  ) {

  }  // iOS: getConversationPage
  func getAllConversations(
    _ resolve: RCTPromiseResolve<[String: Any?]>, rejecter reject: RCTPromiseReject
  ) {

  }  // todo: android use v. large page?
  func getActiveConversations(
    _ resolve: RCTPromiseResolve<[String: Any?]>, rejecter reject: RCTPromiseReject
  ) {

  }  // todo: android can use observed active conv ids, and filter all convs
  func getOpenConversationCount(
    _ resolve: RCTPromiseResolve<Int>, rejecter reject: RCTPromiseReject
  ) {
      reject("\("RNKustomerChatProvider")::NO_IMPL", "Feature not implemented", nil)
  }  // todo: android can use observed active ids count
  func getUnreadCount(_ resolve: RCTPromiseResolve<Int>, rejecter reject: RCTPromiseReject) {
          reject("\("RNKustomerChatProvider")::NO_IMPL", "Feature not implemented", nil)
  }
  // #endregion
}
